crates/bevy_core_pipeline/src/core_3d/mod.rs

```diff
@@ -179,37 +176,48 @@ impl Plugin for Core3dPlugin {
     }
 }
 
+/// Opaque 3D [`BinnedPhaseItem`]s.
 pub struct Opaque3d {
-    pub asset_id: AssetId<Mesh>,
-    pub pipeline: CachedRenderPipelineId,
-    pub entity: Entity,
-    pub draw_function: DrawFunctionId,
+    /// The key, which determines which can be batched.
+    pub key: Opaque3dBinKey,
+    /// An entity from which the mesh and other attributes will be fetched.
+    pub representative_entity: Entity,
+    /// The ranges of instances.
     pub batch_range: Range<u32>,
+    /// The dynamic offset.
     pub dynamic_offset: Option<NonMaxU32>,
 }
 
-impl PhaseItem for Opaque3d {
-    type SortKey = (usize, AssetId<Mesh>);
+/// Data that must be identical in order to batch meshes together.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Opaque3dBinKey {
+    /// The identifier of the render pipeline.
+    pub pipeline: CachedRenderPipelineId,
 
-    #[inline]
-    fn entity(&self) -> Entity {
-        self.entity
-    }
+    /// The function used to draw.
+    pub draw_function: DrawFunctionId,
 
-    #[inline]
-    fn sort_key(&self) -> Self::SortKey {
-        // Sort by pipeline, then by mesh to massively decrease drawcall counts in real scenes.
-        (self.pipeline.id(), self.asset_id)
-    }
+    /// The mesh.
+    pub asset_id: AssetId<Mesh>,
```

Comment:
Thankfully when we move to slotmap render assets this will drop from 16 bytes to 8.